src/wasm/wasm-type.cpp

```diff
@@ -1319,85 +1333,102 @@ void Canonicalizer::scanType(Type* type) {
   }
 }
 
-void Canonicalizer::makeAncestorsFixedPoint() {
-  // Naively calculate the transitive closure of the reachability graph.
+void Canonicalizer::findSelfReferentialHeapTypes() {
+  // Calculate the fixed point of the reachability relation.
+  auto fixedPoint = children;
   bool changed;
   do {
     changed = false;
-    for (auto& entry : ancestors) {
+    for (auto& entry : fixedPoint) {
       auto& succs = entry.second;
-      std::unordered_set<TypeID> nextAncestors;
+      std::unordered_set<TypeID> newSuccs;
       for (auto& other : succs) {
-        auto& otherAncestors = ancestors[other];
-        nextAncestors.insert(otherAncestors.begin(), otherAncestors.end());
+        auto& otherSuccs = fixedPoint[other];
+        newSuccs.insert(otherSuccs.begin(), otherSuccs.end());
       }
       size_t oldSize = succs.size();
-      succs.insert(nextAncestors.begin(), nextAncestors.end());
+      succs.insert(newSuccs.begin(), newSuccs.end());
       if (succs.size() != oldSize) {
         changed = true;
       }
     }
   } while (changed);
+
+  // Find HeapTypes that reach themselves
+  for (auto& entry : builder.impl->entries) {
```

Comment:
Can this loop be on `fixedPoint`? Basically the set we care about is `{ x \in children(x) | x \in AllThings }` so I'm not sure why this starts with `builder.impl->entries`.

Reply:
If we looped over `fixedPoint`, we would have to first traverse it to collect a set of all TypeIDs it references and at some point we would also have to check whether each TypeID corresponded to a HeapType. In contrast, the HeapTypes correspond 1:1 to builder entries, so it takes less code find them that way.

Reply:
I see, sounds good.